<h1 style="text-align: center;">Data Profiling</h1>

## Short explanation
This notebook performs an initial profile of all tables in the football database to understand their structure, size, completeness, uniqueness, value distributions, date coverage, and basic data-quality characteristics. It will also examine key relationships between tables. The findings will later be used to build the project's data catalogue and guide further data exploration and problem definition.

---

### Library Import 

In [1]:
import sqlite3
import pandas as pd

## Database import 

In [6]:
connection = sqlite3.connect(r"..\Data\database.sqlite")

In [8]:
tables = pd.read_sql("SELECT name FROM sqlite_master WHERE type='table';", connection)

In [9]:
tables

,name
0,sqlite_sequence
1,Player_Attributes
2,Player
3,Match
4,League
5,Country
6,Team
7,Team_Attributes


### Data import

In [10]:
country_df = pd.read_sql("SELECT * FROM Country", connection)

In [11]:
country_df.head()

,id,name
0,1,Belgium
1,1729,England
2,4769,France
3,7809,Germany
4,10257,Italy


In [12]:
league_df = pd.read_sql("SELECT * FROM League", connection)

In [13]:
league_df.head()

,id,country_id,name
0,1,1,Belgium Jupiler League
1,1729,1729,England Premier League
2,4769,4769,France Ligue 1
3,7809,7809,Germany 1. Bundesliga
4,10257,10257,Italy Serie A


In [14]:
player_df = pd.read_sql("SELECT * FROM Player", connection)

In [15]:
player_df.head()

,id,player_api_id,player_name,player_fifa_api_id,birthday,height,weight
0,1,505942,Aaron Appindangoye,218353,1992-02-29 00:00:00,182.88,187
1,2,155782,Aaron Cresswell,189615,1989-12-15 00:00:00,170.18,146
2,3,162549,Aaron Doran,186170,1991-05-13 00:00:00,170.18,163
3,4,30572,Aaron Galindo,140161,1982-05-08 00:00:00,182.88,198
4,5,23780,Aaron Hughes,17725,1979-11-08 00:00:00,182.88,154


In [16]:
player_attributes_df = pd.read_sql("SELECT * FROM Player_Attributes", connection)

In [17]:
player_attributes_df.head()

,id,player_fifa_api_id,player_api_id,date,overall_rating,potential,preferred_foot,attacking_work_rate,defensive_work_rate,crossing,...,vision,penalties,marking,standing_tackle,sliding_tackle,gk_diving,gk_handling,gk_kicking,gk_positioning,gk_reflexes
0,1,218353,505942,2016-02-18 00:00:00,67.0,71.0,right,medium,medium,49.0,...,54.0,48.0,65.0,69.0,69.0,6.0,11.0,10.0,8.0,8.0
1,2,218353,505942,2015-11-19 00:00:00,67.0,71.0,right,medium,medium,49.0,...,54.0,48.0,65.0,69.0,69.0,6.0,11.0,10.0,8.0,8.0
2,3,218353,505942,2015-09-21 00:00:00,62.0,66.0,right,medium,medium,49.0,...,54.0,48.0,65.0,66.0,69.0,6.0,11.0,10.0,8.0,8.0
3,4,218353,505942,2015-03-20 00:00:00,61.0,65.0,right,medium,medium,48.0,...,53.0,47.0,62.0,63.0,66.0,5.0,10.0,9.0,7.0,7.0
4,5,218353,505942,2007-02-22 00:00:00,61.0,65.0,right,medium,medium,48.0,...,53.0,47.0,62.0,63.0,66.0,5.0,10.0,9.0,7.0,7.0


In [18]:
team_df = pd.read_sql("SELECT * FROM Team", connection)

In [19]:
team_df.head()

,id,team_api_id,team_fifa_api_id,team_long_name,team_short_name
0,1,9987,673.0,KRC Genk,GEN
1,2,9993,675.0,Beerschot AC,BAC
2,3,10000,15005.0,SV Zulte-Waregem,ZUL
3,4,9994,2007.0,Sporting Lokeren,LOK
4,5,9984,1750.0,KSV Cercle Brugge,CEB


In [22]:
team_attributes_df = pd.read_sql("SELECT * FROM Team_Attributes", connection)

In [23]:
team_attributes_df.head()

,id,team_fifa_api_id,team_api_id,date,buildUpPlaySpeed,buildUpPlaySpeedClass,buildUpPlayDribbling,buildUpPlayDribblingClass,buildUpPlayPassing,buildUpPlayPassingClass,...,chanceCreationShooting,chanceCreationShootingClass,chanceCreationPositioningClass,defencePressure,defencePressureClass,defenceAggression,defenceAggressionClass,defenceTeamWidth,defenceTeamWidthClass,defenceDefenderLineClass
0,1,434,9930,2010-02-22 00:00:00,60,Balanced,NaN,Little,50,Mixed,...,55,Normal,Organised,50,Medium,55,Press,45,Normal,Cover
1,2,434,9930,2014-09-19 00:00:00,52,Balanced,48.0,Normal,56,Mixed,...,64,Normal,Organised,47,Medium,44,Press,54,Normal,Cover
2,3,434,9930,2015-09-10 00:00:00,47,Balanced,41.0,Normal,54,Mixed,...,64,Normal,Organised,47,Medium,44,Press,54,Normal,Cover
3,4,77,8485,2010-02-22 00:00:00,70,Fast,NaN,Little,70,Long,...,70,Lots,Organised,60,Medium,70,Double,70,Wide,Cover
4,5,77,8485,2011-02-22 00:00:00,47,Balanced,NaN,Little,52,Mixed,...,52,Normal,Organised,47,Medium,47,Press,52,Normal,Cover


In [24]:
match_df = pd.read_sql("SELECT * FROM Match", connection)

In [25]:
match_df.head()

,id,country_id,league_id,season,stage,date,match_api_id,home_team_api_id,away_team_api_id,home_team_goal,...,SJA,VCH,VCD,VCA,GBH,GBD,GBA,BSH,BSD,BSA
0,1,1,1,2008/2009,1,2008-08-17 00:00:00,492473,9987,9993,1,...,4.00,1.65,3.40,4.50,1.78,3.25,4.00,1.73,3.40,4.20
1,2,1,1,2008/2009,1,2008-08-16 00:00:00,492474,10000,9994,0,...,3.80,2.00,3.25,3.25,1.85,3.25,3.75,1.91,3.25,3.60
2,3,1,1,2008/2009,1,2008-08-16 00:00:00,492475,9984,8635,0,...,2.50,2.35,3.25,2.65,2.50,3.20,2.50,2.30,3.20,2.75
3,4,1,1,2008/2009,1,2008-08-17 00:00:00,492476,9991,9998,5,...,7.50,1.45,3.75,6.50,1.50,3.75,5.50,1.44,3.75,6.50
4,5,1,1,2008/2009,1,2008-08-16 00:00:00,492477,7947,9985,1,...,1.73,4.50,3.40,1.65,4.50,3.50,1.65,4.75,3.30,1.67


---

## Country Table Profile

In [27]:
country_profile = pd.DataFrame({
    "columns":country_df.columns,
    "Data_type": country_df.dtypes.values,
    "missing_value": country_df.isna().sum().values,
    "missing_percentage": (country_df.isna().mean()*100).values,
    "unique_value":country_df.nunique().values
})

In [28]:
country_profile

,columns,Data_type,missing_value,missing_percentage,unique_value
0,id,int64,0,0.0,11
1,name,str,0,0.0,11


In [29]:
country_df.shape

(11, 2)

In [30]:
country_df.duplicated().sum()

np.int64(0)

In [37]:
country_df

,id,name
0,1,Belgium
1,1729,England
2,4769,France
3,7809,Germany
4,10257,Italy
5,13274,Netherlands
6,15722,Poland
7,17642,Portugal
8,19694,Scotland
9,21518,Spain


### Evaluation 

The Country table is small, containing 11 records and 2 columns. Both columns are complete, with no missing values, and there are no duplicate rows. Each column contains 11 unique values, indicating that both the country IDs and country names are unique within this table. Overall, the table appears structurally clean based on these basic profiling checks.

---

## League Table Profiling

In [39]:
league_profile = pd.DataFrame({
    "Columns":  league_df.columns,
    "Data_type": league_df.dtypes.values,
    "missing_values": league_df.isna().sum().values,
    "missing_percentages": (league_df.isna().mean()*100).values,
    "uniqe_values": league_df.nunique().values
})

In [40]:
league_profile

,Columns,Data_type,missing_values,missing_percentages,uniqe_values
0,id,int64,0,0.0,11
1,country_id,int64,0,0.0,11
2,name,str,0,0.0,11


In [41]:
league_df.shape

(11, 3)

In [42]:
league_df.duplicated().sum()

np.int64(0)

In [43]:
league_df

,id,country_id,name
0,1,1,Belgium Jupiler League
1,1729,1729,England Premier League
2,4769,4769,France Ligue 1
3,7809,7809,Germany 1. Bundesliga
4,10257,10257,Italy Serie A
5,13274,13274,Netherlands Eredivisie
6,15722,15722,Poland Ekstraklasa
7,17642,17642,Portugal Liga ZON Sagres
8,19694,19694,Scotland Premier League
9,21518,21518,Spain LIGA BBVA


### Evaluation
The League table is small, containing 11 records and 3 columns. It has no missing values or duplicate rows, and all three columns contain 11 unique values. The country_id column provides the foreign-key relationship with the Country table, while each league name is unique. Overall, the table appears clean based on the initial profiling checks.

---

## Player Table Profile

In [44]:
player_profile = pd.DataFrame({
    "Columns":player_df.columns,
    "Data_Type": player_df.dtypes.values,
    "Missing_values": player_df.isna().sum().values,
    "Missing_Percentage": (player_df.isna().mean()*100).values,
    "Unique": player_df.nunique().values
})

In [45]:
player_profile

,Columns,Data_Type,Missing_values,Missing_Percentage,Unique
0,id,int64,0,0.0,11060
1,player_api_id,int64,0,0.0,11060
2,player_name,str,0,0.0,10848
3,player_fifa_api_id,int64,0,0.0,11060
4,birthday,str,0,0.0,5762
5,height,float64,0,0.0,20
6,weight,int64,0,0.0,50


In [46]:
player_df.shape

(11060, 7)

In [47]:
player_df.duplicated().sum()

np.int64(0)

In [49]:
player_df.head(10)

,id,player_api_id,player_name,player_fifa_api_id,birthday,height,weight
0,1,505942,Aaron Appindangoye,218353,1992-02-29 00:00:00,182.88,187
1,2,155782,Aaron Cresswell,189615,1989-12-15 00:00:00,170.18,146
2,3,162549,Aaron Doran,186170,1991-05-13 00:00:00,170.18,163
3,4,30572,Aaron Galindo,140161,1982-05-08 00:00:00,182.88,198
4,5,23780,Aaron Hughes,17725,1979-11-08 00:00:00,182.88,154
5,6,27316,Aaron Hunt,158138,1986-09-04 00:00:00,182.88,161
6,7,564793,Aaron Kuhl,221280,1996-01-30 00:00:00,172.72,146
7,8,30895,Aaron Lennon,152747,1987-04-16 00:00:00,165.10,139
8,9,528212,Aaron Lennox,206592,1993-02-19 00:00:00,190.50,181
9,10,101042,Aaron Meijers,188621,1987-10-28 00:00:00,175.26,170


### Evaluation
The Player table contains 11,060 records and 7 columns, with no missing values or duplicate rows. The API and FIFA identifiers are unique, while player names contain some repeated values. Overall, the table appears complete and structurally clean based on the initial profiling checks.

---

## Player Attributes profile

In [50]:
player_attributes_profile = pd.DataFrame({
    "Columns":player_attributes_df.columns,
    "Data_types":player_attributes_df.dtypes.values,
    "Missing_values": player_attributes_df.isna().sum().values,
    "Missing_percentage": (player_attributes_df.isna().mean()*100).values,
    "Unique": player_attributes_df.nunique().values
})

In [51]:
player_attributes_profile

,Columns,Data_types,Missing_values,Missing_percentage,Unique
0,id,int64,0,0.000000,183978
1,player_fifa_api_id,int64,0,0.000000,11062
2,player_api_id,int64,0,0.000000,11060
3,date,str,0,0.000000,197
4,overall_rating,float64,836,0.454402,61
5,potential,float64,836,0.454402,56
6,preferred_foot,str,836,0.454402,2
7,attacking_work_rate,str,3230,1.755645,8
8,defensive_work_rate,str,836,0.454402,19
9,crossing,float64,836,0.454402,95


In [52]:
player_attributes_df.shape

(183978, 42)

In [53]:
player_attributes_df.duplicated().sum()

np.int64(0)

In [54]:
player_attributes_df.head().T

,0,1,2,3,4
id,1,2,3,4,5
player_fifa_api_id,218353,218353,218353,218353,218353
player_api_id,505942,505942,505942,505942,505942
date,2016-02-18 00:00:00,2015-11-19 00:00:00,2015-09-21 00:00:00,2015-03-20 00:00:00,2007-02-22 00:00:00
overall_rating,67.0,67.0,62.0,61.0,61.0
potential,71.0,71.0,66.0,65.0,65.0
preferred_foot,right,right,right,right,right
attacking_work_rate,medium,medium,medium,medium,medium
defensive_work_rate,medium,medium,medium,medium,medium
crossing,49.0,49.0,49.0,48.0,48.0


### Evalutation
The Player_Attributes table contains 183,978 records and 42 columns. No duplicate rows were identified, and the id column is unique. However, several attribute columns contain missing values, with the extent varying across columns. These missing values will be investigated and addressed later during the data-cleaning stage.

---

## Team Table Profile 

In [55]:
team_profile = pd.DataFrame({
    "Columns":team_df.columns,
    "Data_type":team_df.dtypes.values,
    "missing_values":team_df.isna().sum().values,
    "Missing_percentages":(team_df.isna().mean()*100).values,
    "Unique":team_df.nunique().values
})

In [56]:
team_profile

,Columns,Data_type,missing_values,Missing_percentages,Unique
0,id,int64,0,0.00000,299
1,team_api_id,int64,0,0.00000,299
2,team_fifa_api_id,float64,11,3.67893,285
3,team_long_name,str,0,0.00000,296
4,team_short_name,str,0,0.00000,259


In [57]:
team_df.shape

(299, 5)

In [58]:
team_df.duplicated().sum()

np.int64(0)

In [60]:
team_df.head(10)

,id,team_api_id,team_fifa_api_id,team_long_name,team_short_name
0,1,9987,673.0,KRC Genk,GEN
1,2,9993,675.0,Beerschot AC,BAC
2,3,10000,15005.0,SV Zulte-Waregem,ZUL
3,4,9994,2007.0,Sporting Lokeren,LOK
4,5,9984,1750.0,KSV Cercle Brugge,CEB
5,6,8635,229.0,RSC Anderlecht,AND
6,7,9991,674.0,KAA Gent,GEN
7,8,9998,1747.0,RAEC Mons,MON
8,9,7947,NaN,FCV Dender EH,DEN
9,10,9985,232.0,Standard de Liège,STL


### Evaluation 
The Team table contains 299 records and 5 columns, with no duplicate rows. Most fields are complete, although team_fifa_api_id contains 11 missing values. The API identifiers and team names are not all unique, which is worth keeping in mind for later relationship and data-quality checks.

---

## Team Attributes Profile

In [61]:
team_attributes_profile = pd.DataFrame({
    "Columns":team_attributes_df.columns,
    "Data_type":team_attributes_df.dtypes.values,
    "missing_values":team_attributes_df.isna().sum().values,
    "Missing_percentages":(team_attributes_df.isna().mean()*100).values,
    "Unique":team_attributes_df.nunique().values
})

In [62]:
team_attributes_profile

,Columns,Data_type,missing_values,Missing_percentages,Unique
0,id,int64,0,0.000000,1458
1,team_fifa_api_id,int64,0,0.000000,285
2,team_api_id,int64,0,0.000000,288
3,date,str,0,0.000000,6
4,buildUpPlaySpeed,int64,0,0.000000,57
5,buildUpPlaySpeedClass,str,0,0.000000,3
6,buildUpPlayDribbling,float64,969,66.460905,49
7,buildUpPlayDribblingClass,str,0,0.000000,3
8,buildUpPlayPassing,int64,0,0.000000,58
9,buildUpPlayPassingClass,str,0,0.000000,3


In [63]:
team_attributes_df.shape

(1458, 25)

In [64]:
team_attributes_df.duplicated().sum()

np.int64(0)

In [66]:
team_attributes_df.head(5).T

,0,1,2,3,4
id,1,2,3,4,5
team_fifa_api_id,434,434,434,77,77
team_api_id,9930,9930,9930,8485,8485
date,2010-02-22 00:00:00,2014-09-19 00:00:00,2015-09-10 00:00:00,2010-02-22 00:00:00,2011-02-22 00:00:00
buildUpPlaySpeed,60,52,47,70,47
buildUpPlaySpeedClass,Balanced,Balanced,Balanced,Fast,Balanced
buildUpPlayDribbling,NaN,48.0,41.0,NaN,NaN
buildUpPlayDribblingClass,Little,Normal,Normal,Little,Little
buildUpPlayPassing,50,56,54,70,52
buildUpPlayPassingClass,Mixed,Mixed,Mixed,Long,Mixed


### Evaluation
The Team_Attributes table contains 1,458 records and 25 columns, with no duplicate rows. Most columns are complete, with buildUpPlayDribbling having approximately 66% missing values. The remaining columns contain no missing values. The id column is unique, while the other identifiers contain repeated values.

---

## Match Profile

In [67]:
match_profile = pd.DataFrame({
    "Columns":match_df.columns,
    "Data_type":match_df.dtypes.values,
    "missing_values":match_df.isna().sum().values,
    "Missing_percentages":(match_df.isna().mean()*100).values,
    "Unique":match_df.nunique().values
})

In [71]:
pd.set_option("display.max_rows",None)
pd.set_option("display.max_columns",None)

match_profile

,Columns,Data_type,missing_values,Missing_percentages,Unique
0,id,int64,0,0.000000,25979
1,country_id,int64,0,0.000000,11
2,league_id,int64,0,0.000000,11
3,season,str,0,0.000000,8
4,stage,int64,0,0.000000,38
5,date,str,0,0.000000,1694
6,match_api_id,int64,0,0.000000,25979
7,home_team_api_id,int64,0,0.000000,299
8,away_team_api_id,int64,0,0.000000,299
9,home_team_goal,int64,0,0.000000,11


In [69]:
match_df.shape

(25979, 115)

In [70]:
match_df.duplicated().sum()

np.int64(0)

### Evaluation 
The remaining fields mainly represent pre-match betting odds from different bookmakers for home win, draw, and away win outcomes. Several bookmaker groups have substantial missing values, with some fields missing for around 45% of matches. Importantly, the Match table does not contain an explicit classification target, so a target may need to be derived from the available match-result information. This will be considered later when defining the ML problem.

---

# Profiling Summary

Initial profiling of all seven database tables has been completed. The main structural characteristics, missing values, duplicates, and key observations have been documented.

Detailed table metadata and profiling findings are available in:

`docs/data_catalog/`

The original field and abbreviation reference is available in:

`docs/football_data_reference_notes.txt`

The data catalogue should be used as the project-specific reference for the database, while the Football-Data reference notes should be used to interpret documented field abbreviations and betting-related variables.

Further data-quality checks, relationship validation, and target definition will be carried out in the next stages of the project.